src/lib/rust.ts

```diff
@@ -328,130 +317,178 @@ const tryDownloadingBinary = async ({
+  for (const [name, version] of await resolveMarineAndMreplDependencies()) {
     // Not installing dependencies in parallel
     // for cargo logs to be clearly readable
     // eslint-disable-next-line no-await-in-loop
-    await ensureCargoDependency({
-      nameAndVersion: `${name}@${version}`,
-      maybeFluenceConfig,
-      force,
-    });
+    await ensureMarineOrMreplDependency({ name, version });
   }
-};
+}
 
 type ResolveVersionArg = {
   name: string;
-  maybeVersion: string | undefined;
-  maybeFluenceConfig: FluenceConfig | null;
+  version: string | undefined;
 };
 
 async function resolveVersionToInstall({
   name,
-  maybeVersion,
-  maybeFluenceConfig,
+  version,
 }: ResolveVersionArg): Promise<string | undefined> {
-  if (typeof maybeVersion === "string") {
-    return (await isExactVersion(maybeVersion)) ? maybeVersion : undefined;
+  if (typeof version === "string") {
+    return (await isExactVersion(version)) ? version : undefined;
   }
 
-  const maybeRecommendedVersion = isFluenceCargoDependency(name)
-    ? versions.cargo[name]
+  const fluenceConfig = await initFluenceConfig();
+  return isMarineOrMrepl(name)
+    ? fluenceConfig?.[`${name}Version`] ?? versions.cargo[name]
     : undefined;
+}
+
+export async function resolveMarineAndMreplDependencies() {
+  const fluenceConfig = await initFluenceConfig();
+
+  return [
+    [
+      "marine",
+      fluenceConfig?.marineVersion === undefined
+        ? versions.cargo.marine
+        : fluenceConfig.marineVersion,
+    ],
+    [
+      "mrepl",
+      fluenceConfig?.mreplVersion === undefined
+        ? versions.cargo.mrepl
+        : fluenceConfig.mreplVersion,
+    ],
+  ] as const;
+}
+
+type ResolveDependencyDirPathAndTmpPath = {
+  name: string;
+  version: string;
+};
 
-  const maybeKnownVersion =
-    maybeFluenceConfig?.dependencies?.cargo?.[name] ?? maybeRecommendedVersion;
+export async function resolveDependencyDirPathAndTmpPath({
+  name,
+  version,
+}: ResolveDependencyDirPathAndTmpPath): Promise<{
+  dependencyTmpDirPath: string;
+  dependencyDirPath: string;
+}> {
+  const [depDirPath, depTmpDirPath] = await Promise.all([
+    ensureUserFluenceCargoDir(),
+    ensureUserFluenceTmpCargoDir(),
+  ]);
+
+  const dependencyPathEnding = join(...name.split("/"), version);
+  return {
+    dependencyTmpDirPath: join(depTmpDirPath, dependencyPathEnding),
+    dependencyDirPath: join(depDirPath, dependencyPathEnding),
+  };
+}
 
-  return maybeKnownVersion;
+type HandleInstallationArg = {
+  dependencyDirPath: string;
+  dependencyTmpDirPath: string;
+  name: string;
+  version: string;
+  toolchain: string | undefined;
+};
+
+export async function installUsingCargo({
+  dependencyDirPath,
+  dependencyTmpDirPath,
+  name,
+  version,
+  toolchain,
+}: HandleInstallationArg): Promise<void> {
+  try {
+    // if dependency is already installed it will be there
+    // so there is no need to install
+    await access(dependencyDirPath);
```

Comment:
Do you need add `return` statement here if dependency is installed and that's it ?

Reply:
I think no real reason to add it cause this function doesn't contain anything else anyway and will exit on it's own
Only if some error happens - then the code in the catch will run

Reply:
Yes, you're right, sorry. Just misread the code.